plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.ClusterStateUpdateTask;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.metadata.QueryGroup;
+import org.opensearch.cluster.service.ClusterManagerTaskThrottler.ThrottlingKey;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.Priority;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.rest.RestStatus;
+import org.opensearch.plugin.wlm.action.CreateQueryGroupResponse;
+import org.opensearch.search.ResourceType;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static org.opensearch.search.query_group.QueryGroupServiceSettings.MAX_QUERY_GROUP_COUNT;
+import static org.opensearch.search.query_group.QueryGroupServiceSettings.QUERY_GROUP_COUNT_SETTING_NAME;
+
+/**
+ * This class defines the functions for QueryGroup persistence
+ */
+public class QueryGroupPersistenceService {
+    private static final Logger logger = LogManager.getLogger(QueryGroupPersistenceService.class);
+    private final ClusterService clusterService;
+    private static final String SOURCE = "query-group-persistence-service";
+    private static final String CREATE_QUERY_GROUP_THROTTLING_KEY = "create-query-group";
+    private static final String UPDATE_QUERY_GROUP_THROTTLING_KEY = "update-query-group";
+    private static final String DELETE_QUERY_GROUP_THROTTLING_KEY = "delete-query-group";
+    private volatile int maxQueryGroupCount;
+    final ThrottlingKey createQueryGroupThrottlingKey;
+    final ThrottlingKey updateQueryGroupThrottlingKey;
+    final ThrottlingKey deleteQueryGroupThrottlingKey;
+
+    /**
+     * Constructor for QueryGroupPersistenceService
+     *
+     * @param clusterService {@link ClusterService} - The cluster service to be used by QueryGroupPersistenceService
+     * @param settings {@link Settings} - The settings to be used by QueryGroupPersistenceService
+     * @param clusterSettings {@link ClusterSettings} - The cluster settings to be used by QueryGroupPersistenceService
+     */
+    @Inject
+    public QueryGroupPersistenceService(
+        final ClusterService clusterService,
+        final Settings settings,
+        final ClusterSettings clusterSettings
+    ) {
+        this.clusterService = clusterService;
+        this.createQueryGroupThrottlingKey = clusterService.registerClusterManagerTask(CREATE_QUERY_GROUP_THROTTLING_KEY, true);
+        this.deleteQueryGroupThrottlingKey = clusterService.registerClusterManagerTask(DELETE_QUERY_GROUP_THROTTLING_KEY, true);
+        this.updateQueryGroupThrottlingKey = clusterService.registerClusterManagerTask(UPDATE_QUERY_GROUP_THROTTLING_KEY, true);
+        maxQueryGroupCount = MAX_QUERY_GROUP_COUNT.get(settings);
```

Comment:
There is no validation when the setting is applied for the first time, if specified in yml? check https://github.com/opensearch-project/OpenSearch/issues/14777